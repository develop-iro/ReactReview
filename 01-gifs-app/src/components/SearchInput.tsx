import { useState, useEffect } from "react";
interface SearchInputProps {
  buttonText?: string;
  placeHolder: string;
  ariaLabel?: string;
  onSearch?: (value: string) => void;
}

export const SearchInput = ({
  buttonText,
  placeHolder,
  ariaLabel,
  onSearch,
}: SearchInputProps) => {
  const [query, setQuery] = useState("");

  useEffect(() => {
    const timer = setTimeout(() => {
      onSearch?.(query);
    }, 800);

    return () => clearTimeout(timer);
  }, [query, onSearch]);

  const handleSearch = () => {
    onSearch?.(query);
  };
  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter") {
      handleSearch();
    }
  };

  return (
    <div className="search-container content-center">
      <input
        type="text"
        placeholder={placeHolder}
        aria-label={ariaLabel}
        value={query}
        onChange={(e) => setQuery(e.target.value)}
        onKeyDown={handleKeyDown}
        onBlur={handleSearch}
      />
      <button onClick={handleSearch}>{buttonText}</button>
    </div>
  );
};
