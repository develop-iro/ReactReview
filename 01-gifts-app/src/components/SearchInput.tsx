interface SearchInputProps {
  placeHolder: string;
  ariaLabel?: string;
}

export const SearchInput = ({ placeHolder, ariaLabel }: SearchInputProps) => {
  return (
    <div className="search-container content-center">
      <input type="text" placeholder={placeHolder} aria-label={ariaLabel} />
    </div>
  );
};
