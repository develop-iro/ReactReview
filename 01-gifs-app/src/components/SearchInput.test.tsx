import { describe, it, expect, vi } from "vitest";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { SearchInput } from "./SearchInput";

describe("SearchInput", () => {
  it("should render correctly", () => {
    const { container } = render(<SearchInput placeHolder={""} />);
    expect(container).toMatchSnapshot();
  });

  it("should call correct onSearch function after 800ms", async () => {
    const fn = vi.fn();
    render(<SearchInput placeHolder={""} onSearch={fn} />);
    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "matrix" } });

    await waitFor(() => {
      expect(fn).toHaveBeenCalled();
      expect(fn).toHaveBeenCalledWith("matrix");
    });
  });

  it("should only once with last value after 800ms", async () => {
    const fn = vi.fn();
    render(<SearchInput placeHolder={""} onSearch={fn} />);
    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "m" } });
    fireEvent.change(input, { target: { value: "ma" } });
    fireEvent.change(input, { target: { value: "mat" } });
    fireEvent.change(input, { target: { value: "matr" } });
    fireEvent.change(input, { target: { value: "matri" } });
    fireEvent.change(input, { target: { value: "matrix" } });
    await waitFor(() => {
      expect(fn).toHaveBeenCalledTimes(1);
      expect(fn).toHaveBeenCalledWith("matrix");
    });
  });
  it("should call onSearch on button click", () => {
    const fn = vi.fn();
    render(<SearchInput placeHolder={""} onSearch={fn} />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(fn).toHaveBeenCalled();
  });

  it("should call onSearch on input blur", () => {
    const fn = vi.fn();
    render(<SearchInput placeHolder={""} onSearch={fn} />);
    const input = screen.getByRole("textbox");
    fireEvent.blur(input);
    expect(fn).toHaveBeenCalled();
  });
  it("should call onSearch on Enter key press", () => {
    const fn = vi.fn();
    render(<SearchInput placeHolder={""} onSearch={fn} />);
    const input = screen.getByRole("textbox");
    fireEvent.keyDown(input, { key: "Enter" });
    expect(fn).toHaveBeenCalled();
  });
  it("should set placeholder text", () => {
    const { getByRole } = render(<SearchInput placeHolder={"Search..."} />);
    const input = getByRole("textbox");
    expect(input.hasAttribute("placeholder")).toBe(true);
    expect(input.getAttribute("placeholder")).toBe("Search...");
  });
});
