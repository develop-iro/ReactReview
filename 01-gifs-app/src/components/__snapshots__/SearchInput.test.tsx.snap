// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchInput > should render correctly 1`] = `
<div>
  <div
    class="search-container content-center"
  >
    <input
      placeholder=""
      type="text"
      value=""
    />
    <button />
  </div>
</div>
`;
